mujoco_py/mjpid.pyx

```diff
@@ -0,0 +1,122 @@
+    cdef double dt_in_sec = m.opt.timestep
+
+    cdef double error = d.ctrl[id] - d.actuator_length[id]
+
+    cdef double Kp = m.actuator_user[id * m.nuser_actuator + PROPORTIONAL_GAIN]
+    cdef double error_deadband = m.actuator_user[id * m.nuser_actuator + ERROR_DEADBAND]
+    cdef double integral_max_clamp = m.actuator_user[id * m.nuser_actuator + INTEGRAL_MAX_CLAMP]
+    cdef double integral_time_const = m.actuator_user[id * m.nuser_actuator + INTEGRAL_TIME_CONSTANT]
+    cdef double derivative_gain_smoothing = \
+        m.actuator_user[id * m.nuser_actuator + DERIVATIVE_GAIN_SMOOTHING]
+    cdef double derivate_time_const = m.actuator_user[id * m.nuser_actuator + DERIVATIVE_TIME_CONSTANT]
+
+    cdef double corrective_effort_limit = fmax(
+            fabs(m.actuator_forcerange[id * 2]), fabs(m.actuator_forcerange[id * 2 + 1]))
+
+    if fabs(error) < error_deadband:
+        error = 0.0
+
+    integral_error = d.userdata[id * NUM_USER_DATA_PER_ACT + INTEGRAL_ERROR]
+    integral_error += error * dt_in_sec
+    integral_error = fmax(-integral_max_clamp, fmin(integral_max_clamp, integral_error))
+
+    last_error = d.userdata[id * NUM_USER_DATA_PER_ACT + LAST_ERROR]
+    cdef double derivative_error = (error - last_error) / dt_in_sec
+
+    derivative_error_last = d.userdata[id * NUM_USER_DATA_PER_ACT + DERIVATIVE_ERROR_LAST]
+
+    derivative_error = (1.0 - derivative_gain_smoothing) * derivative_error_last + \
+        derivative_gain_smoothing * derivative_error
+
+    cdef double integral_error_term = 0.0
+    if integral_time_const != 0:
+        integral_error_term = integral_error / integral_time_const
+
+    cdef double derivative_error_term = 0.0
+    if derivate_time_const != 0:
+        derivative_error_term = derivative_error / derivate_time_const
+
+    f = Kp * (error + integral_error_term + derivative_error_term)
+    # print(id, error, integral_error_term, derivative_error_term, derivative_error, dt_in_sec,
+    #    last_error, integral_error, derivative_error_last)
+
+    d.userdata[id * NUM_USER_DATA_PER_ACT + LAST_ERROR] = error
+    d.userdata[id * NUM_USER_DATA_PER_ACT + DERIVATIVE_ERROR_LAST] = derivative_error
+    d.userdata[id * NUM_USER_DATA_PER_ACT + INTEGRAL_ERROR] = integral_error
+    return fmax(-corrective_effort_limit, fmin(corrective_effort_limit, f))
+
+
+def set_pid_control(m, d):
+    global mjcb_act_gain
+    global mjcb_act_bias
+
+    if m.nuserdata < m.nu * NUM_USER_DATA_PER_ACT:
+        raise Exception('nuserdata is not set large enough to store PID internal states')
+
+    for i in range(m.nuserdata):
+        d.userdata[i] = 0.0
+
+    for i in range(m.nu):
+        m.actuator_gaintype[i] = const.GAIN_USER
+        m.actuator_biastype[i] = const.BIAS_USER
+
+        # if user does not set, will chose default settings.
+        # for kp, it tries to use m.actuator_gainprm[i][0]
+        if m.actuator_user[i][PROPORTIONAL_GAIN] <= 0.0:
```

Comment:
I know it's not that important, but why not use gainprm for this purpose?

Reply:
I also see this kind of as a good thing. If we keep the `kp` at the same position, just by changing the `gaintype` from `user` to `position`, we can change between the P and PID controller with the same P value. Pretty cool, isn't it?